XML to RCF Editor2/functions.py

```diff
@@ -0,0 +1,207 @@
+dictionary (json file) or not, if found, edit messages values inside enforcement, if not continue looping till finding the
+following rule tag and so on.
+inputs: xml_file, dictionary
+outputs: None
+"""
+
+def loop_and_enable(element, json_dictionary):
+    # loop on parents
+    for sub in element:
+
+        # check if current parent is rule or not
+        if sub.tag == "rule":
+
+            # if it's rule, check if it exists in json dictionary or not
+            rule_id = sub.get("id")
+            if rule_id in json_dictionary:
+
+                ## if found in the dictionary, get the messages IDs to be edited
+                # activate this rule then enable messages
+                element.set('active', 'yes')
+                messages_id = json_dictionary[rule_id]
+                enable_messages(sub,messages_id)
+
+
+        # if parent has children, loop on them
+        if len(sub) > 1:
+                loop_and_enable(sub, json_dictionary)
+
+"""
+function name: edit_mapped
+function description: function taking rule parent, searching for enforcement inside, then search for messages ids in the
+given list, and enable them
+inputs: element (rule), messages_key_val (list containing messages to be enabled)
+return: None (editing the xml)
+"""
+
+def enable_messages(element, messages_key_val, tag = "enforcement" ,message_key = "id", key_to_change = "mapped", key_to_change_val = "yes"):
+
+
+    # search for tag (enforcement) inside element (rule)
+    enforcement = element.find(f'{tag}')
+
+    # check if tag (enforcement) exists or not
+    if enforcement is None: # if not found print tag not found
+        print(f"{tag} not found")
+
+    else: # if tag (enforcement) found, search inside for the messages id and enable them
+
+        # loop on the given messages ids
+        for msg in messages_key_val:
+
+            message = enforcement.find(f".//*[@{message_key}='{msg}']")
+
+            # check if the message exists or not
+            if message is not None: # if found, set key_to_change (mapped) to key_to_change_val ("yes")
+                message.set(key_to_change, key_to_change_val)
+                print(f"message with id = {msg} under rule id {element.get("id")} found and modified")
+
+            else: # if not found, add it
+
+                # copy the first message
+                first_message = enforcement[0] if len(enforcement) > 0 else None
+                # get a clone
+                new_message = copy.deepcopy(first_message)
+                # edit the clone attributes: set id, then set the required attribute
+                new_message.set(message_key, msg)
+                new_message.set(key_to_change, key_to_change_val)
+                # append the clone to the XML file
+                enforcement.append(new_message)
+                print(f"message with id = {msg} under rule id {element.get("id")} does not exist. message added")
+
+
+
+"""
+function name: find_and_edit
```

Comment:
why are we having this function?

Reply:
this function was for testing. function removed from functions.py and added to test.py